packages/ipfs-http-client/package.json

```diff
@@ -20,13 +20,6 @@
     "go-ipfs": false,
     "ipfs-core-utils/src/files/normalise-input": "ipfs-core-utils/src/files/normalise-input/index.browser.js"
   },
-  "typesVersions": {
```

Comment:
Without this libraries importing http client won’t get any types

Reply:
I thought it might break something.  Thing is, with this running the typecheck on the ipfs-cli package fails with:

```
ipfs-cli: npm info it worked if it ends with ok
ipfs-cli: npm info using npm@6.14.1
ipfs-cli: npm info using node@v12.16.1
ipfs-cli: npm info lifecycle ipfs-cli@0.0.1~prebuild: ipfs-cli@0.0.1
ipfs-cli: npm info lifecycle ipfs-cli@0.0.1~build: ipfs-cli@0.0.1
ipfs-cli: > ipfs-cli@0.0.1 build /Users/alex/Documents/Workspaces/ipfs/js-ipfs/packages/ipfs-cli
ipfs-cli: > tsc --emitDeclarationOnly --declarationDir dist
ipfs-cli: src/daemon.js(10,32): error TS2307: Cannot find module 'ipfs-http-client' or its corresponding type declarations.
ipfs-cli: src/utils.js(130,26): error TS2307: Cannot find module 'ipfs-http-client' or its corresponding type declarations.
ipfs-cli: npm info lifecycle ipfs-cli@0.0.1~build: Failed to exec build script
ipfs-cli: npm ERR! code ELIFECYCLE
ipfs-cli: npm ERR! errno 1
ipfs-cli: npm ERR! ipfs-cli@0.0.1 build: `tsc --emitDeclarationOnly --declarationDir dist`
ipfs-cli: npm ERR! Exit status 1
ipfs-cli: npm ERR! 
ipfs-cli: npm ERR! Failed at the ipfs-cli@0.0.1 build script.
ipfs-cli: npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
ipfs-cli: npm timing npm Completed in 4477ms
```

Running `tsc` with `--traceResolution` shows it's prefixing every path with `./dist` when trying to load `ipfs-http-client`:

```
======== Resolving module 'ipfs-http-client' from '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/packages/ipfs-cli/src/daemon.js'. ========
Explicitly specified module resolution kind: 'NodeJs'.
Loading module 'ipfs-http-client' from 'node_modules' folder, target file type 'TypeScript'.
Directory '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/packages/ipfs-cli/src/node_modules' does not exist, skipping all lookups in it.
Found 'package.json' at '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/packages/ipfs-cli/node_modules/ipfs-http-client/package.json'.
'package.json' has a 'typesVersions' field with version-specific path mappings.
File '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/packages/ipfs-cli/node_modules/ipfs-http-client.ts' does not exist.
File '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/packages/ipfs-cli/node_modules/ipfs-http-client.tsx' does not exist.
File '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/packages/ipfs-cli/node_modules/ipfs-http-client.d.ts' does not exist.
'package.json' does not have a 'typings' field.
'package.json' does not have a 'types' field.
'package.json' has 'main' field 'src/index.js' that references '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/packages/ipfs-cli/node_modules/ipfs-http-client/src/index.js'.
'package.json' has a 'typesVersions' entry '*' that matches compiler version '4.0.3', looking for a pattern to match module name 'src/index.js'.
Module name 'src/index.js', matched pattern '*'.
Trying substitution 'dist/*', candidate module location: 'dist/src/index.js'.
File '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/packages/ipfs-cli/node_modules/ipfs-http-client/dist/src/index.js' does not exist.
Loading module as file / folder, candidate module location '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/packages/ipfs-cli/node_modules/ipfs-http-client/dist/src/index.js', target file type 'TypeScript'.
File name '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/packages/ipfs-cli/node_modules/ipfs-http-client/dist/src/index.js' has a '.js' extension - stripping it.
Directory '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/packages/ipfs-cli/node_modules/@types' does not exist, skipping all lookups in it.
Directory '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/packages/node_modules' does not exist, skipping all lookups in it.
File '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/node_modules/ipfs-http-client.ts' does not exist.
File '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/node_modules/ipfs-http-client.tsx' does not exist.
File '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/node_modules/ipfs-http-client.d.ts' does not exist.
File '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/node_modules/@types/ipfs-http-client.d.ts' does not exist.
Directory '/Users/alex/Documents/Workspaces/ipfs/node_modules' does not exist, skipping all lookups in it.
Directory '/Users/alex/Documents/Workspaces/node_modules' does not exist, skipping all lookups in it.
Directory '/Users/alex/Documents/node_modules' does not exist, skipping all lookups in it.
Directory '/Users/alex/node_modules' does not exist, skipping all lookups in it.
Directory '/Users/node_modules' does not exist, skipping all lookups in it.
Directory '/node_modules' does not exist, skipping all lookups in it.
Loading module 'ipfs-http-client' from 'node_modules' folder, target file type 'JavaScript'.
Directory '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/packages/ipfs-cli/src/node_modules' does not exist, skipping all lookups in it.
Found 'package.json' at '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/packages/ipfs-cli/node_modules/ipfs-http-client/package.json'.
'package.json' has a 'typesVersions' field with version-specific path mappings.
File '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/packages/ipfs-cli/node_modules/ipfs-http-client.js' does not exist.
File '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/packages/ipfs-cli/node_modules/ipfs-http-client.jsx' does not exist.
'package.json' has 'main' field 'src/index.js' that references '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/packages/ipfs-cli/node_modules/ipfs-http-client/src/index.js'.
'package.json' has a 'typesVersions' entry '*' that matches compiler version '4.0.3', looking for a pattern to match module name 'src/index.js'.
Module name 'src/index.js', matched pattern '*'.
Trying substitution 'dist/*', candidate module location: 'dist/src/index.js'.
File '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/packages/ipfs-cli/node_modules/ipfs-http-client/dist/src/index.js' does not exist.
Loading module as file / folder, candidate module location '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/packages/ipfs-cli/node_modules/ipfs-http-client/dist/src/index.js', target file type 'JavaScript'.
File name '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/packages/ipfs-cli/node_modules/ipfs-http-client/dist/src/index.js' has a '.js' extension - stripping it.
Directory '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/packages/node_modules' does not exist, skipping all lookups in it.
File '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/node_modules/ipfs-http-client.js' does not exist.
File '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/node_modules/ipfs-http-client.jsx' does not exist.
Directory '/Users/alex/Documents/Workspaces/ipfs/node_modules' does not exist, skipping all lookups in it.
Directory '/Users/alex/Documents/Workspaces/node_modules' does not exist, skipping all lookups in it.
Directory '/Users/alex/Documents/node_modules' does not exist, skipping all lookups in it.
Directory '/Users/alex/node_modules' does not exist, skipping all lookups in it.
Directory '/Users/node_modules' does not exist, skipping all lookups in it.
Directory '/node_modules' does not exist, skipping all lookups in it.
Loading module 'ipfs-http-client' from 'node_modules' folder, target file type 'Json'.
Directory '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/packages/ipfs-cli/src/node_modules' does not exist, skipping all lookups in it.
Found 'package.json' at '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/packages/ipfs-cli/node_modules/ipfs-http-client/package.json'.
'package.json' has a 'typesVersions' field with version-specific path mappings.
'package.json' has 'main' field 'src/index.js' that references '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/packages/ipfs-cli/node_modules/ipfs-http-client/src/index.js'.
'package.json' has a 'typesVersions' entry '*' that matches compiler version '4.0.3', looking for a pattern to match module name 'src/index.js'.
Module name 'src/index.js', matched pattern '*'.
Trying substitution 'dist/*', candidate module location: 'dist/src/index.js'.
File '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/packages/ipfs-cli/node_modules/ipfs-http-client/dist/src/index.js' does not exist.
Loading module as file / folder, candidate module location '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/packages/ipfs-cli/node_modules/ipfs-http-client/dist/src/index.js', target file type 'Json'.
Directory '/Users/alex/Documents/Workspaces/ipfs/js-ipfs/packages/node_modules' does not exist, skipping all lookups in it.
Directory '/Users/alex/Documents/Workspaces/ipfs/node_modules' does not exist, skipping all lookups in it.
Directory '/Users/alex/Documents/Workspaces/node_modules' does not exist, skipping all lookups in it.
Directory '/Users/alex/Documents/node_modules' does not exist, skipping all lookups in it.
Directory '/Users/alex/node_modules' does not exist, skipping all lookups in it.
Directory '/Users/node_modules' does not exist, skipping all lookups in it.
Directory '/node_modules' does not exist, skipping all lookups in it.
======== Module name 'ipfs-http-client' was not resolved. ========
```

Any ideas how to get it to not do that?

Reply:
It fails to find probably because types for ipfs-http-client had not been generated when you run that. And it should look in `dist` because that’s what typeVersions does. It maps TS search path to the dist folder.

Unfortunately as things stand today ts will not pickup type information from js files if they come from the different package, e.g. ipfs-core will not be able to get type information from ipfs-core-utils even though both have jsdoc comments. It is very unfortunate & I hope TS team will fix it at some point, but as things stand today we need to generate types for them to get picked across package boundaries.

Reply:
No, the types are there in `/packages/ipfs-http-client/dist`.